be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala

```diff
@@ -0,0 +1,112 @@
+package ch.epfl.pop.authentication
+
+import akka.http.scaladsl.model.{AttributeKey, HttpRequest, HttpResponse, StatusCodes}
+import akka.http.scaladsl.server
+import akka.http.scaladsl.server.Directives._
+
+/** Helper object used to build a server route for handling authentication requests
+  */
+object Authenticate {
+  private case class RequestParameters(response_type: String, client_id: String, redirect_uri: String, scope: String, state: Option[String], response_mode: Option[String], login_hint: String, nonce: String)
+
+  private type VerificationState = Either[(String, String), Unit]
+
+  private val mandatoryParameters = Set("response_type", "client_id", "redirect_uri", "scope", "login_hint", "nonce")
+
+  /** Builds a route that can handle an authentication request
+    * @return
+    *   A route either successfully handling a given authentication request or rejecting it with an error
+    */
+  def buildRoute(): server.Route = {
+    extractRequest { request =>
+
+      val validParametersRoute = parameters(
+        "response_type",
+        "client_id",
+        "redirect_uri",
+        "scope",
+        "state".optional,
+        "response_mode".optional,
+        "login_hint",
+        "nonce"
+      ) {
+        (response_type, client_id, redirect_uri, scope, state, response_mode, login_hint, nonce) =>
+          {
+            val params = RequestParameters(response_type, client_id, redirect_uri, scope, state, response_mode, login_hint, nonce)
+            verifyParameters(params) match {
+              case Left(error -> errorDescription) => complete(authenticationFailure(error, errorDescription, state))
+              case Right(_)                        => complete(generateChallenge(request))
+            }
+          }
+      }
+
+      val invalidParametersRoute = extractUri { uri =>
+        complete {
+          val parametersFound = uri.query().toMap.keys.toSet
+          val missingParams = mandatoryParameters.diff(parametersFound)
+          val errorDescription = missingParams.map(name => s"[$name]").mkString("Missing parameters: ", " ", "")
+          authenticationFailure("invalid_request", errorDescription, None)
+        }
+      }
+
+      validParametersRoute ~ invalidParametersRoute
+    }
+  }
+
+  private def verifyParameters(params: RequestParameters): VerificationState = {
+    for {
+      _ <- verifyResponseType(params.response_type)
+      _ <- verifyScope(params.scope)
+      _ <- verifyRedirectUri(params.redirect_uri)
+      result <- verifyResponseMode(params.response_mode)
+    } yield result
+  }
+
+  private def verifyResponseType(response_type: String): VerificationState = {
+    val expectedResponseType = "id_token token"
+    if (response_type == expectedResponseType)
+      Right(())
+    else
+      Left("unsupported_response_type" -> s"expected \"$expectedResponseType\" but received \"$response_type\"")
```

Comment:
It is a general comment: I suggest that you declare the type of errors as constant values. 
You can then reuse them in your test suites.

Reply:
That makes sense! I put them directly in the Authenticate object as they are not enough general-purpose to justify a separate file for them (and there are only 3).